src/Subsampling/include/gudhi/choose_n_farthest_points.h

```diff
@@ -142,6 +146,217 @@ void choose_n_farthest_points(Distance dist,
+ *  \brief Subsample by a greedy strategy of iteratively adding the farthest point from the
+ *  current chosen point set to the subsampling.
+ *  \details
+ *  This computes the same thing as `choose_n_farthest_points()`, but relies on the triangle
+ *  inequality to reduce the amount of computation when the doubling dimension and spread are small.
+ *  In the worst case, this can be much slower than `choose_n_farthest_points()` though.
+ *  See \cite sheehy20onehop and its references for details about this algorithm.
+ *  \tparam Distance must provide an operator() that takes 2 points (value type of the range)
+ *  and returns their distance as a `double`. It must be a true metric, the algorithm relies on the triangle inequality.
+ *  \tparam Point_range Random access range of points.
+ *  \tparam PointOutputIterator Output iterator whose value type is the point type.
+ *  \tparam DistanceOutputIterator Output iterator for distances.
+ * @param[in] dist A distance function.
+ * @param[in] input_pts The input points.
+ * @param[in] final_size The size of the subsample to compute.
+ * @param[in] starting_point The seed in the farthest point algorithm.
+ * @param[out] output_it The output iterator for points.
+ * @param[out] dist_it The optional output iterator for distances.
+ */
+template < typename Distance,
+typename Point_range,
+typename PointOutputIterator,
+typename DistanceOutputIterator = Null_output_iterator>
+void choose_n_farthest_points_metric(Distance dist_,
+                                     Point_range const &input_pts,
+                                     std::size_t final_size,
+                                     std::size_t starting_point,
+                                     PointOutputIterator output_it,
+                                     DistanceOutputIterator dist_it = {}) {
+  std::size_t nb_points = boost::size(input_pts);
+  if (final_size > nb_points)
+    final_size = nb_points;
+
+  // Tests to the limit
+  if (final_size < 1)
+    return;
+
+  if (starting_point == random_starting_point) {
+    // Choose randomly the first landmark
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    std::uniform_int_distribution<std::size_t> dis(0, nb_points - 1);
+    starting_point = dis(gen);
+  }
+
+  // FIXME: don't hard-code the type as double. For Epeck_d, we also want to handle types that do not have an infinity.
+  typedef double FT;
+  static_assert(std::numeric_limits<FT>::has_infinity, "the number type needs to support infinity()");
+
+  *output_it++ = input_pts[starting_point];
+  *dist_it++ = std::numeric_limits<FT>::infinity();
+  if (final_size == 1) return;
+
+  auto dist = [&](std::size_t a, std::size_t b){ return dist_(input_pts[a], input_pts[b]); };
+
+  std::vector<Landmark_info<FT>> landmarks(nb_points);
+  radius_priority_ds<FT> radius_priority(&landmarks);
+
+  auto compute_radius = [&](std::size_t i)
+  {
+    FT r = -std::numeric_limits<FT>::infinity(); // -2 * diameter should suffice
```

Comment:
Is it possible that a value in dist_ is negative ? Values in landmarks[i].voronoi are only initialized with dist(), so I am wondering why r needs to be that small.

Reply:
Distances cannot be negative, and we could use 0 here. However there is at least one place in the code where I would normally want to check if the cell is empty or too far. With a radius of -infinity, empty cells are detected as too far and we can skip the explicit test for emptiness. This is a small optimization.

Reply:
Ok, I see. It is just that it forces FT to have an infinity, which seems to be a problem from one of the comments above (l.224: "For Epeck_d, we also want to handle types that do not have an infinity.").

Reply:
Yes, the comment is there to give a hint what to do if we ever decide to remove this FT restriction, but it would have been good if I had marked more clearly which test uses this optimization :disappointed: